controllers/syncer/call_home.go

```diff
@@ -0,0 +1,168 @@
+ * limitations under the License.
+ */
+
+package syncer
+
+import (
+	"github.com/imdario/mergo"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/IBM/ibm-block-csi-operator/controllers/internal/ibmblockcsi"
+	"github.com/IBM/ibm-block-csi-operator/pkg/config"
+	"github.com/IBM/ibm-block-csi-operator/pkg/util/boolptr"
+	"github.com/presslabs/controller-util/mergo/transformers"
+	"github.com/presslabs/controller-util/syncer"
+)
+
+const (
+	callHomeContainerName = "ibm-block-csi-call-home"
+	secretVolumeName      = "call-home-secret-dir"
+	// CronSchedule runs once a day at midnight
+	CronSchedule     = "0 0 * * *"
+	jobsHistoryLimit = int32(1)
+)
+
+type callHomeSyncer struct {
+	driver *ibmblockcsi.IBMBlockCSI
+	obj    runtime.Object
+}
+
+// NewCallHomeSyncer returns a syncer for call home
+func NewCallHomeSyncer(c client.Client, scheme *runtime.Scheme, driver *ibmblockcsi.IBMBlockCSI) syncer.Interface {
+	obj := &batchv1.CronJob{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        config.GetNameForResource(config.CallHome, driver.Name),
+			Namespace:   driver.Namespace,
+			Annotations: driver.GetAnnotations("", ""),
+			Labels:      driver.GetLabels(),
+		},
+	}
+
+	sync := &callHomeSyncer{
+		driver: driver,
+		obj:    obj,
+	}
+
+	return syncer.NewObjectSyncer(config.CallHome.String(), driver.Unwrap(), obj, c, func() error {
+		return sync.SyncFn()
+	})
+}
+
+func (s *callHomeSyncer) SyncFn() error {
+	out := s.obj.(*batchv1.CronJob)
+
+	out.Spec.Schedule = CronSchedule
+
+	// ensure template
+	out.Spec.JobTemplate.ObjectMeta.Labels = s.driver.GetCallHomePodLabels()
+	out.Spec.JobTemplate.ObjectMeta.Annotations = s.driver.GetAnnotations("", "")
+	out.Spec.JobTemplate.Spec.Template.ObjectMeta.Labels = s.driver.GetCallHomePodLabels()
+	successfulJobsHistoryLimit := jobsHistoryLimit
+	out.Spec.SuccessfulJobsHistoryLimit = &successfulJobsHistoryLimit
+
+	err := mergo.Merge(&out.Spec.JobTemplate.Spec.Template.Spec, s.ensurePodSpec(), mergo.WithTransformers(transformers.PodSpec))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *callHomeSyncer) ensurePodSpec() corev1.PodSpec {
+	return corev1.PodSpec{
+		Containers:         s.ensureContainersSpec(),
+		Volumes:            s.ensureVolumes(),
+		ServiceAccountName: config.GetNameForResource(config.CallHomeServiceAccount, s.driver.Name),
+		Affinity:           s.driver.Spec.CallHome.Affinity,
+		Tolerations:        s.driver.Spec.CallHome.Tolerations,
+		RestartPolicy:      "OnFailure",
+	}
+}
+
+func (s *callHomeSyncer) ensureContainersSpec() []corev1.Container {
+	callHomePlugin := s.ensureContainer(callHomeContainerName,
+		s.driver.GetCallHomeImage(),
+		[]string{"--csi-endpoint=$(CSI_ENDPOINT)"},
+	)
+
+	callHomePlugin.Resources = ensureResources("40m", "800m", "40Mi", "400Mi")
```

Comment:
where did we get these numbers?

Reply:
I'm not sure the controller needs the same amount of resources as call home

Reply:
CSI-3953

Reply:
could you open a ticket to calibrate it?